lib/Dist/Zilla/Dist/Builder.pm

```diff
@@ -462,11 +462,28 @@ might not exist.
 
 sub archive_filename {
   my ($self) = @_;
-  return join(q{},
+
+  my $name = join(q{},
     $self->dist_basename,
     ( $self->is_trial && $self->version !~ /_/ ? '-TRIAL' : '' ),
-    '.tar.gz'
   );
+
+  my $provided_filename;
+
+  for my $namer ($self->plugins_with(-ArchiveNamer)->@*) {
+    my $this_filename = $namer->archive_filename($name);
+
+    next unless defined $this_filename;
+
+    $self->log_fatal('attempted to set archive_filename twice')
+      if defined $provided_filename;
+
+    $provided_filename = $this_filename;
```

Comment:
When you do this, you should also capture the plugin name.  Then, when you hit the log_fatal condition, you can say which two plugins are responsible.